python/sglang/srt/layers/quantization/rtn_utils.py

```diff
@@ -0,0 +1,127 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+# Copyright © 2025, Oracle and/or its affiliates.
+
+import logging
+
+import torch
+from torch.nn.parameter import Parameter
+
+logger = logging.getLogger(__name__)
+
+
+def rtn_quantize(
+    tensor: torch.Tensor, num_bits: int, group_size: int
+) -> tuple[torch.Tensor, torch.Tensor]:
+    """Quantize a tensor using per-group static scaling factor.
+
+    Args:
+        tensor: The input tensor.
+        num_bits: Target precision for the result (supported values are
+                  8 or 4).
+        group_size: Quantization granularity.
+                    If equal to -1, each row in the input tensor is treated
+                    as one group.
+    """
+    batch_present = len(tensor.shape) == 3
+    if not batch_present:
+        tensor = tensor.unsqueeze(0)
+
+    q_range = 2**num_bits
+    num_groups = (
+        tensor.shape[1] * tensor.shape[2] // group_size
+        if group_size != -1
+        else tensor.shape[1]
+    )
+    # Calculate a scaling factor per input group.
+    input_flat = tensor.reshape(tensor.shape[0], num_groups, -1)
+    input_min = torch.min(input_flat, dim=2, keepdim=True)[0]
+    input_max = torch.max(input_flat, dim=2, keepdim=True)[0]
+    input_max_abs = torch.max(input_min.abs(), input_max.abs())
+    scale = input_max_abs * 2.0 / (q_range - 1)
+    # Scale each input group, truncate and round to the nearest integer.
```

Comment:
This comment should read `Scale each input group, round to the nearest integer, shift the range and truncate`